source/common/quic/quic_server_transport_socket_factory.cc

```diff
@@ -26,11 +27,88 @@ QuicServerTransportSocketConfigFactory::createTransportSocketFactory(
   return factory;
 }
 
+namespace {
+void initializeQuicCertAndKey(Ssl::TlsContext& context,
+                              const Ssl::TlsCertificateConfig& /*cert_config*/) {
+  // Convert the certificate chain loaded into the context into PEM, as that is what the QUICHE
+  // API expects. By using the version already loaded, instead of loading it from the source,
+  // we can reuse all the code that loads from different formats, allows using passwords on the key,
+  // etc.
+  std::vector<std::string> chain;
+  auto process_one_cert = [&](X509* cert) {
+    bssl::UniquePtr<BIO> bio(BIO_new(BIO_s_mem()));
+    int result = PEM_write_bio_X509(bio.get(), cert);
+    ASSERT(result == 1);
+    BUF_MEM* buf_mem = nullptr;
+    result = BIO_get_mem_ptr(bio.get(), &buf_mem);
+    std::string cert_str(buf_mem->data, buf_mem->length);
+    std::istringstream pem_stream(cert_str);
+    auto pem_result = quic::ReadNextPemMessage(&pem_stream);
+    RELEASE_ASSERT(pem_result.status == quic::PemReadResult::Status::kOk,
+                   "Failed to read already loaded cert");
+    chain.push_back(std::move(pem_result.contents));
+  };
+
+  process_one_cert(SSL_CTX_get0_certificate(context.ssl_ctx_.get()));
+
+  STACK_OF(X509)* chain_stack = nullptr;
+  int result = SSL_CTX_get0_chain_certs(context.ssl_ctx_.get(), &chain_stack);
+  ASSERT(result == 1);
+  for (size_t i = 0; i < sk_X509_num(chain_stack); i++) {
+    process_one_cert(sk_X509_value(chain_stack, i));
+  }
+
+  quiche::QuicheReferenceCountedPointer<quic::ProofSource::Chain> cert_chain(
+      new quic::ProofSource::Chain(chain));
+
+  std::string error_details;
+  bssl::UniquePtr<EVP_PKEY> pub_key(X509_get_pubkey(context.cert_chain_.get()));
+  int sign_alg = deduceSignatureAlgorithmFromPublicKey(pub_key.get(), &error_details);
+  if (sign_alg == 0) {
+    throwEnvoyExceptionOrPanic(
+        absl::StrCat("Failed to deduce signature algorithm from public key: ", error_details));
+  }
+
+  context.quic_cert_ = std::move(cert_chain);
+
+  bssl::UniquePtr<EVP_PKEY> privateKey(
+      bssl::UpRef(SSL_CTX_get0_privatekey(context.ssl_ctx_.get())));
+  std::unique_ptr<quic::CertificatePrivateKey> pem_key =
+      std::make_unique<quic::CertificatePrivateKey>(std::move(privateKey));
+  if (pem_key == nullptr) {
+    throwEnvoyExceptionOrPanic("Failed to load QUIC private key.");
+  }
+
+  context.quic_private_key_ = std::move(pem_key);
+}
+} // namespace
+
+QuicServerTransportSocketFactory::QuicServerTransportSocketFactory(
+    bool enable_early_data, Stats::Scope& scope, Ssl::ServerContextConfigPtr config,
+    Envoy::Ssl::ContextManager& manager, const std::vector<std::string>& server_names)
+    : QuicTransportSocketFactoryBase(scope, "server"), manager_(manager), stats_scope_(scope),
+      config_(std::move(config)), server_names_(server_names),
+      ssl_ctx_(Runtime::runtimeFeatureEnabled(
+                   "envoy.restart_features.quic_handle_certs_with_shared_tls_code")
```

Comment:
There's still going to be some inconsistency between the proof source and the transport socket factory, and that's not easily fixable, because the proof source has already committed to a codepath before it can resolve the filter chain so that it can then find the correct transport socket (they're not 1:1). 

I experimented with latching the value at process startup to truly make it require a restart to change, but that prevented testing it.

But I'll latch here to prevent a stale `ssl_ctx_` from being used if the flag is disabled at runtime.

But really, you shouldn't do that to a restart-feature flag.